src/track/track.cpp

```diff
@@ -204,6 +204,12 @@ QString Track::getCanonicalLocation() const {
     // (copy-on write). But operating on a single instance of QFileInfo
     // might not be thread-safe due to internal caching!
     QMutexLocker lock(&m_qMutex);
+    if (!m_fileInfo.exists()) {
+        // don't use the cached empty canonical location, if the track was not
+        // seen before.
+        // The user might have resored the track in the meanwhile.
+        m_fileInfo.refresh();
```

Comment:
I propose to split getFileInfo() into cachedFileInfo() and freshFileInfo() to explicitly distinguish both use cases:
```
const QFileInfo& Track::cachedFileInfo() const {
   // QFileInfo is implicitly shared and thread-safe, no locking needed.
   DEBUG_ASSERT(m_fileInfo.caching());
   return m_fileInfo;
}

const QFileInfo& Track::freshFileInfo() const {
   // QFileInfo is implicitly shared and thread-safe, no locking needed.
   // ...add some notes why we only refresh for (yet) non-existent files
   // and ignore the potential disappearance of files...
   DEBUG_ASSERT(m_fileInfo.caching());
   if (!m_fileInfo.exists() {
      m_fileInfo.refresh();
   }
   return m_fileInfo;
}

bool Track::exists() const {
   return freshFileInfo().exists()
}
```

It is not  needed to acquire the mutex in freshFileInfo(), because QFileInfo is implicitly shared and will be detached as soon as the first thread calls QFileInfo::refresh(). Due to race conditions QFileInfo::refresh() might be invoked multiple times, but that doesn't matter. The result of invoking refresh() depends the file system as an external resource, so a mutex does not provide any benefits and only gives a false sense of determinism.

After thinking about this I also came to the conclusion that it is safe to return a const-ref to the m_fileInfo member.

We need to check where m_fileInfo and getFileInfo() is used and decide which of those two functions to call!